### ISSUE TITLE
Valahlla prefers longer trivial route when destination within a traffic closure

### ISSUE THREAD

**Issue by mandeepsandhu** [valhalla/valhalla#2709]
Consider a trivial map like this:
```
          E----F
          |    |
      A---B----C-2-D
```

When routing from `A -> 2`, it takes the path `AB`,  `BC` & `CD`.

However if we close the edge `CD` due to traffic, then for the same route, it starts preferring the longer `BEFC` rather than `BC`.

**Comment by mandeepsandhu** [comment]
Test case demonstrating the issue:
```c++
namespace {
  inline void SetLiveSpeed(baldr::TrafficSpeed* live_speed, uint64_t speed) {
    live_speed->breakpoint1 = 255;
    live_speed->overall_speed = speed >> 1;
    live_speed->speed1 = speed >> 1;
  }

  void close_single_edge(baldr::GraphReader& reader,
                         baldr::TrafficTile& tile,
                         int index,
                         baldr::TrafficSpeed* current,
                         const std::string& edge_name,
                         const gurka::map& closure_map) {
    baldr::GraphId tile_id(tile.header->tile_id);
    std::string start_node(1, edge_name.front());
    std::string end_node(1, edge_name.back());

    auto fwd = std::get<0>(gurka::findEdge(reader, closure_map.nodes, edge_name, start_node));
    auto back = std::get<0>(gurka::findEdge(reader, closure_map.nodes, edge_name, end_node));
    bool should_close = (fwd.Tile_Base() == tile_id && fwd.id() == index) ||
                        (back.Tile_Base() == tile_id && back.id() == index);
    if (should_close) {
      SetLiveSpeed(current, 0);
    }
  }
  } // namespace

class ClosureAtDest : public ::testing::TestWithParam<std::string> {
  protected:
    static gurka::map closure_map;
    static int const default_speed;
    static std::string const tile_dir;
    static std::shared_ptr<baldr::GraphReader> reader;
    ;

    static void SetUpTestSuite() {
      const std::string ascii_map = R"(
          E----F
          |    |
      A---B----C-2-D

    )";

      const std::string speed_str = std::to_string(default_speed);
      const gurka::ways ways = {{"AB", {{"highway", "primary"}, {"maxspeed", speed_str}}},
                                {"BC", {{"highway", "primary"}, {"maxspeed", speed_str}}},
                                {"CD", {{"highway", "primary"}, {"maxspeed", speed_str}}},
                                {"BEFC", {{"highway", "residential"}, {"maxspeed", "10"}}}};

      const auto layout = gurka::detail::map_to_coordinates(ascii_map, 10, {.05f, .2f});
      closure_map = gurka::buildtiles(layout, ways, {}, {}, tile_dir);

      closure_map.config.put("mjolnir.traffic_extract", tile_dir + "/traffic.tar");
      test::build_live_traffic_data(closure_map.config);

      reader = test::make_clean_graphreader(closure_map.config.get_child("mjolnir"));
    }

    void set_default_speed_on_all_edges() {
      test::customize_live_traffic_data(closure_map.config,
                                        [](baldr::GraphReader& reader, baldr::TrafficTile& tile,
                                           int index, baldr::TrafficSpeed* current) -> void {
                                          SetLiveSpeed(current, default_speed);
                                        });
    }

  virtual void SetUp() {
      set_default_speed_on_all_edges();
    }

    virtual void TearDown() {
      set_default_speed_on_all_edges();
    }
  };

gurka::map ClosureAtDest::closure_map = {};
const int ClosureAtDest::default_speed = 36;
const std::string ClosureAtDest::tile_dir =
    "test/data/closure_at_dest";
std::shared_ptr<baldr::GraphReader> ClosureAtDest::reader;

TEST_P(ClosureAtDest, Test) {
    std::string costing = GetParam();
    {
      auto result =
          gurka::route(closure_map, {"A", "2"}, costing, {{"/date_time/type", "0"}}, reader);
      gurka::assert::osrm::expect_steps(result, {"AB"});
      gurka::assert::raw::expect_path(result, {"AB", "BC", "CD"});
    }
    // Close destination edge
    {
      test::LiveTrafficCustomize close_edge = [](baldr::GraphReader& reader, baldr::TrafficTile& tile,
                                           int index, baldr::TrafficSpeed* current) -> void {
        close_single_edge(reader, tile, index, current, "CD", closure_map);
      };
      test::customize_live_traffic_data(closure_map.config, close_edge);

      auto result =
          gurka::route(closure_map, {"A", "2"}, costing, {{"/date_time/type", "0"}}, reader);
      gurka::assert::osrm::expect_steps(result, {"AB"});
      gurka::assert::raw::expect_path(result, {"AB", "BC", "CD"});
    }
  }
```

**Comment by kevinkreiser** [comment]
In investigating this a little further with @mandeepsandhu it turns out to be all about wrong assumptions on how snapping works or should work.

`2` is far enough away from `C` that it will never node snap to it. Node snapping works by allow the input coordinates to be within some tolerance of a node in the graph. In this case the input coordinate `2` is well outside of that tolerance. However this case seems to illustrate a problem with that type of behavior. If the actual snap point is basically on the end of the edge but not actually snapped to the node then the router only gets the one pair of edges that connected to that node and not the other edges at that node. this can lead to a sort of strange behavior where the router is forced to take a particular route because its the only route that is possible. we could consider not basing the snap tolerance on the original input location but on the snapped location with respect to the node but this will have a lot of changes which we should first vet. i vaguely remember on-purpose not doing this as i was tempted to do it before.

the other interesting thing here is that the correlated edges do indeed have a distance along set to 0/1, ie they are node snapped. that casuses the path finding to discard the outbound edge from the destination, which dissallows tthe most efficient path.


so basically 2 things are happening.

1. loki says the input location cannot be node snapped as its too far away so it pics whatever is the closest edge. but because there are multiple, its the luck of the draw which one you get
2. the distance along the edges are still set to 0/1 which tells thor it is a node snap. thor then removes the 0 percent along edge for the destination and has to take the around the block path to reach the 100 percent along edge.
